diff --git a/scripts/setlocalversion b/scripts/setlocalversion
index 3d3babac8298..6837f96adc3b 100755
--- a/scripts/setlocalversion
+++ b/scripts/setlocalversion
@@ -156,15 +156,6 @@ config_localversion=$(sed -n 's/^CONFIG_LOCALVERSION=\(.*\)$/\1/p' include/confi
 if grep -q "^CONFIG_LOCALVERSION_AUTO=y$" include/config/auto.conf; then
	# full scm version string
	scm_version="$(scm_version)"
-elif [ "${LOCALVERSION+set}" != "set" ]; then
-	# If the variable LOCALVERSION is not set, append a plus
-	# sign if the repository is not in a clean annotated or
-	# signed tagged state (as git describe only looks at signed
-	# or annotated tags - git tag -a/-s).
-	#
-	# If the variable LOCALVERSION is set (including being set
-	# to an empty string), we don't want to append a plus sign.
-	scm_version="$(scm_version --short)"
 fi
 
 echo "${KERNELVERSION}${file_localversion}${config_localversion}${LOCALVERSION}${scm_version}"
