diff --git a/scripts/setlocalversion b/scripts/setlocalversion
index 6b54e46a0f12..b8cad33d0016 100755
--- a/scripts/setlocalversion
+++ b/scripts/setlocalversion
@@ -140,7 +140,6 @@ elif [ "${LOCALVERSION+set}" != "set" ]; then
 	# If the variable LOCALVERSION is set (including being set
 	# to an empty string), we don't want to append a plus sign.
 	scm=$(scm_version --short)
-	res="$res${scm:++}"
 fi
 
 echo "$res"
